hydra-cluster/test/Test/EndToEndSpec.hs

```diff
@@ -357,15 +358,43 @@ spec = around showLogsOnFailure $ do
           version `shouldSatisfy` (=~ ("[0-9]+\\.[0-9]+\\.[0-9]+(-[a-zA-Z0-9]+)?" :: String))
 
       it "logs its command line arguments" $ \_ -> do
-        failAfter 60 $
+        failAfter 10 $
           withTempDir "temp-dir-to-check-hydra-logs" $ \dir -> do
             let hydraSK = dir </> "hydra.sk"
             hydraSKey :: SigningKey HydraKey <- generate arbitrary
             void $ writeFileTextEnvelope hydraSK Nothing hydraSKey
             withCreateProcess (proc "hydra-node" ["-n", "hydra-node-1", "--hydra-signing-key", hydraSK]){std_out = CreatePipe} $
-              \_ (Just nodeOutput) _ _ -> do
-                out <- hGetLine nodeOutput
-                out ^? key "message" . key "tag" `shouldBe` Just (Aeson.String "NodeOptions")
+              \_ (Just nodeOutput) _ _ ->
+                assertALogMatches
+                  nodeOutput
+                  100
+                  (\line -> line ^? key "message" . key "tag" == Just (Aeson.String "NodeOptions"))
+                  "JSON object with key NodeOptions"
+
+assertALogMatches :: Handle -> Int -> (String -> Bool) -> String -> IO ()
+assertALogMatches nodeOutput nbLines predicate failureMessage =
+  go nbLines []
+ where
+  go :: Int -> [String] -> IO ()
+  go 0 messages =
+    fail messages
+  go n messages = do
+    out <- tryJust (guard . isEOFError) $ hGetLine nodeOutput
+    case out of
+      Left _eofError -> fail messages
+      Right out' ->
+        if predicate out'
+          then return ()
+          else go (n - 1) $ out' : messages
+  fail :: [String] -> IO ()
+  fail messages =
+    failure $
+      "No log message matching the predicate found in the "
+        <> show nbLines
+        <> " first lines of logs: "
+        <> failureMessage
+        <> "\nCaptured logs:\n"
+        <> show messages
```

Comment:
This does not print the messages when the `failAfter` times out after 10 seconds.

So instead of retrying 100 times (`nbLines`), I think we should have `assertALogMatches` to also time out after some time (and remove the outer one if we want).

For that to work, you likely need to write a `TVar` with the messages and retrieve it's content when the timeout (on blocking `hGetLine`) triggered.